File: python/design_browser_history.py
class BrowserHistory:

    def __init__(self, homepage: str):
        self.history = [homepage]
        self.ptr = 0 

    def visit(self, url: str) -> None:
        self.ptr += 1
        self.history = self.history[:self.ptr]
        self.history.append(url)
               
    def back(self, steps: int) -> str:
        self.ptr = max(0, self.ptr - steps)
        return self.history[self.ptr]

    def forward(self, steps: int) -> str:
        self.ptr = min(len(self.history) - 1, self.ptr + steps)
        return self.history[self.ptr]    

File: src/DesignBrowserHistory.java
class BrowserHistory {
    private List<String> history;
    private int ptr;

    public BrowserHistory(String homepage) {
        this.history = new ArrayList<>();
        this.history.add(homepage);
        this.ptr = 0;
    }

    public void visit(String url) {
        this.ptr++;
        this.history = this.history.subList(0, this.ptr);
        this.history.add(url);
    }

    public String back(int steps) {
        this.ptr = Math.max(0, this.ptr - steps);
        return this.history.get(this.ptr);
    }

    public String forward(int steps) {
        this.ptr = Math.min(this.history.size() - 1, this.ptr + steps);
        return this.history.get(this.ptr);
    }
}
